src/main/java/jackyy/dimensionaledibles/islands/IslandManager.java

```diff
@@ -0,0 +1,136 @@
+import net.minecraft.world.storage.MapStorage;
+import net.minecraft.world.storage.WorldSavedData;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.UUID;
+
+import static jackyy.dimensionaledibles.DimensionalEdibles.MODID;
+import static jackyy.dimensionaledibles.DimensionalEdibles.logger;
+
+public class IslandManager {
+    private static final HashMap<World, IslandManager> MANAGERS = new HashMap<>();
+    private final IslandsWorldSavedData data;
+
+    private IslandManager(World world) {
+        data = IslandsWorldSavedData.get(world);
+    }
+
+    public static IslandManager forWorld(World world) {
+        IslandManager manager = MANAGERS.get(world);
+        if (manager == null) {
+            manager = new IslandManager(world);
+            MANAGERS.put(world, manager);
+        }
+        return manager;
+    }
+
+    /**
+     * Returns an island for a player by id, creating one if the player doesn't already have an island
+     */
+    public Island getIslandForPlayer(UUID id) {
+        Island island = data.getByOwningPlayer(id);
+        if (island == null) {
+            island = data.createPlayerOwnedIsland(id);
+        }
+        return island;
+    }
+
+    /**
+     * Returns an island for a team by id, creating one if the team doesn't already have an island
+     */
+    public Island getIslandForTeam(short id) {
+        Island island = data.getByOwningTeam(id);
+        if (island == null) {
+            island = data.createTeamOwnedIsland(id);
+        }
+        return island;
+    }
+
+    public static class IslandsWorldSavedData extends WorldSavedData {
+        private static final int COMPOUND_TYPE = 10;
+        private static final String DATA_NAME = MODID + "_IslandData";
+        private final HashMap<UUID, Island> islands = new HashMap<>();
+        private int index = 0;
+
+        public IslandsWorldSavedData() {
+            super(DATA_NAME);
+        }
+
+        @SuppressWarnings("unused")
+        public IslandsWorldSavedData(String s) {
+            super(s);
+        }
+
+        public static IslandsWorldSavedData get(World world) {
+            MapStorage storage = world.getPerWorldStorage();
+            IslandsWorldSavedData instance = (IslandsWorldSavedData) storage.getOrLoadData(IslandsWorldSavedData.class, DATA_NAME);
+
+            if (instance == null) {
+                instance = new IslandsWorldSavedData();
+                storage.setData(DATA_NAME, instance);
+            }
+            return instance;
+        }
+
+        @Nullable
+        public Island getByOwningPlayer(UUID id) {
+            return islands.values().stream().filter(i -> i.getOwningPlayer().equals(id)).findFirst().orElse(null);
+        }
+
+        @Nullable
+        public Island getByOwningTeam(short id) {
+            return islands.values().stream().filter(i -> i.getOwningTeam() == id).findFirst().orElse(null);
```

Comment:
Since there is only ever 1 island per team, `findFirst()` is not really needed here.

Reply:
How would you change it? findFirst is still needed since we are filtering a stream from what I can tell.

Reply:
Streams require a terminal operation be called somewhere, otherwise nothing happens; they are lazily evaluated.

He wants a single result value. `findFirst()` is a terminal operation which returns an Optional monad, which either has a retrievable value or indicates the absence of a value. It will however throw an exception if the encountered value is `null`.

What Caedis is doing here is getting that Optional and returning either its encapsulated value, or `null` if there was no result.